lib/Parse/ParseExpr.cpp

```diff
@@ -1386,6 +1386,55 @@ ExprResult Parser::ParseCastExpression(bool isUnaryExpression,
   // are compiling for OpenCL, we need to return an error as this implies
   // that the address of the function is being taken, which is illegal in CL.
 
+  // With primary-expression, before parsing postfix-expression, check to make
+  // sure that Expr is declRefExpr of generic function with _For_any specifier
+  Expr* resExpr = Res.get();
+  if (resExpr != NULL && isa<DeclRefExpr>(resExpr)) {
+    DeclRefExpr* declRef = dyn_cast<DeclRefExpr>(resExpr);
+    ValueDecl* resDecl = declRef->getDecl();
+    if (resDecl != NULL && isa<FunctionDecl>(resDecl)) {
+      FunctionDecl* funDecl = dyn_cast<FunctionDecl>(resDecl);
+      if (funDecl->IsGenericFunction()) {
+        // In this case, parse for <types, ...> in generic function call.
+        ExpectAndConsume(tok::less);
+        bool breakOutOfWhile = false;
+        bool lastEncounteredIsType = false;
+        while (!breakOutOfWhile) {
+          switch (Tok.getKind()) {
+          case tok::comma :
+          case tok::greater :
+            if (!lastEncounteredIsType) {
+              Diag(Tok, diag::err_generic_function_unexpected_token);
```

Comment:
You need to add some error recovery code to this parsing code.   It might easier of the code had the structure:
1. Parse the less than.
2. See if the next token is greater than.  If it is, stop parsing.
3. If the next token is not greater than, parse the list of type names.  Repeat until greater than is reached or an error occurs:
- Parse type name.
- See if the next token is a ',' or '>'.   If it is not, report an error.
- If it is a comma, consume it.

If there is an error, we should at least try to consume all tokens until a 'greater than' or a ';" is hit.   If you look at ParseBoundsExpression, you'll see some error recovery code.